### PR TITLE
Feature: Categories

### PR DESCRIPTION
**Pull request by caseykhuc**
## Reference
- [Design doc](https://coda.io/d/_d4yUiqwZ3dd/Design-document_suPx2)

## Coverage
image

### CODE REVIEW

File: main/controllers/categories.py [hunk 1]
```diff
@@ -0,0 +1,68 @@
...
+    return category
+
+
+@router.get(
+    "/categories/{category_id}",
+    response_model=CategorySchema,
+)
+async def _get_category(category_id: int):
+    category = await get_category_or_404(category_id)
+    return category
```
Comment (kientranngoc) [inline main/controllers/categories.py]: Category is not necessary in this case. We can `return await get_category_or_404(category_id)` directly

File: main/controllers/categories.py [hunk 2]
```diff
@@ -0,0 +1,68 @@
...
+
+    return category
+
+
+@router.get(
+    "/categories/{category_id}",
+    response_model=CategorySchema,
+)
+async def _get_category(category_id: int):
+    category = await get_category_or_404(category_id)
```
Comment (kientranngoc) [inline main/controllers/categories.py]: Should validate the category_id before making a call to db

File: main/controllers/categories.py [hunk 3]
```diff
@@ -0,0 +1,68 @@
...
+)
+async def _add_category(
+    category: CategoryCreatePayloadSchema,
+    user_id: Annotated[int, Depends(require_authentication)],
+):
+    await validate_category_name(category.name)
+    category = await add_category(name=category.name, creator_id=user_id)
+
+    if not category:
+        raise NotFound()
```
Comment (kientranngoc) [inline main/controllers/categories.py]: How can this statement be reached?

File: main/controllers/categories.py [hunk 4]
```diff
@@ -0,0 +1,68 @@
...
+
+@router.delete(
+    "/categories/{category_id}",
+    response_model=Empty,
+    dependencies=[Depends(require_category_creator)],
+)
+async def _delete_category(
+    category_id: int,
+):
+    await delete_category(category_id)
```
Comment (kientranngoc) [inline main/controllers/categories.py:59]: Should validate the category_id before making a call to db

File: main/engines/categories.py [hunk 5]
```diff
@@ -0,0 +1,50 @@
...
+    name: str,
+    creator_id: int,
+) -> CategoryModel:
+    category = CategoryModel(name=name, creator_id=creator_id)
+
+    try:
+        db.session.add(category)
+        await db.session.commit()
+    except SQLAlchemyError:
+        raise BadRequest()
```
Comment (kientranngoc) [inline main/engines/categories.py]: BaseError is specific to return to the client and should only be raised in controllers. If you need to raise an exception from engines, create a custom exception class. Btw, there are a lot of places where you catch SQLAlchemyError and raise BadRequest. Should we do it in one place?

File: main/utils/category.py [hunk 6]
```diff
@@ -0,0 +1,49 @@
...
+
+    if not category:
+        raise NotFound()
+
+    return category
+
+
+async def require_category_creator(
+    category_id: int,
+    id: Annotated[int, Depends(require_authentication)],
```
Comment (ducminh-phan) [inline main/utils/category.py]: The variable name `id` is too generic, it's also a built-in function. Let's rename it to `user_id`

File: tests/controllers/test_category.py [hunk 7]
```diff
@@ -0,0 +1,189 @@
...
+
+async def get_access_token(client, user: UserModel):
+    return (
+        (
+            await client.post(
+                "/login",
+                json={"email": user.email, "password": mock_password},
+            )
+        ).json()
+    )["access_token"]
```
Comment (ducminh-phan) [inline tests/controllers/test_category.py]: We can simply call `create_access_token_from_id` directly

File: main/controllers/categories.py (hunk 2 above)
Comment (caseykhuc) [reply main/controllers/categories.py]: defining param `category_id: int` already handle validation. 
current validation error response:
image

is there another validation you're suggesting to add?

File: main/engines/categories.py (hunk 5 above)
Comment (caseykhuc) [reply main/engines/categories.py]: Updated to handle `SQLAlchemyError` in `main/commons/error_handlers.py`.
This error should only be raised by edge-case scenarios: race conditions - 2 workers adding category with the same name to DB simultaneously for example. Therefore I'd go the a simple solution here.

File: tests/controllers/test_category.py (hunk 7 above)
Comment (caseykhuc) [reply tests/controllers/test_category.py]: Fixed

File: main/utils/category.py (hunk 6 above)
Comment (caseykhuc) [reply main/utils/category.py]: Fixed

File: main/controllers/categories.py (hunk 3 above)
Comment (caseykhuc) [reply main/controllers/categories.py]: Fixed

File: main/controllers/categories.py (hunk 1 above)
Comment (caseykhuc) [reply main/controllers/categories.py]: Fixed

File: main/controllers/categories.py (hunk 4 above)
Comment (caseykhuc) [reply main/controllers/categories.py:59]: Similar to https://github.com/caseykhuc/fastapi-service-learning/pull/2#discussion_r1339501458

File: main/controllers/categories.py (hunk 2 above)
Comment (caseykhuc) [reply main/controllers/categories.py]: Added positive int validation for category_id.

File: main/controllers/categories.py (hunk 4 above)
Comment (caseykhuc) [reply main/controllers/categories.py:59]: Added positive int validation for category_id.

File: main/engines/categories.py (hunk 5 above)
Comment (caseykhuc) [reply main/engines/categories.py]: Removed, use 500 Internal server error for these cases.

**Review by kientranngoc — APPROVED** [review]

**Review by ducminh-phan — APPROVED** [review]
